### ISSUE TITLE
Unique type IDs for comparing types at runtime

### ISSUE THREAD

**Issue by czgdp1807** [lfortran/lfortran#1389]
So, we need type IDs which are unique in order to compare types at runtime and executing the code block associated with those types. This is needed for select-type feature (implemented in #1346) and dynamic binding of functions.

GFortran compares types by comparing the address of their virtual tables. Taking the code snippet from GFortran's IR in https://github.com/lfortran/lfortran/pull/1346#issue-1596382685,

```c
D.3235 = (integer(kind=8)D.4) bptrD.3227._vptrD.3225;
D.3236 = (integer(kind=8)D.4) &__vtab_select_type_01_BaseD.3213;
if (D.3235 == D.3236)
.
.
.
D.3239 = (integer(kind=8)D.4) &__vtab_select_type_01_ChildD.3222;
if (D.3235 == D.3239)
```

That's one of the ways we can use. However for that we will have to create virtual tables globally with `void*` type and then initialise them inside a function and call it before any other function. The drawback of this approach that virtual tables are created for a function even if that function doesn't use runtime polymorphism which is redundant. We should create virtual tables only in those scopes where runtime polymorphism is actually used. 

One more way is to use any of the following approaches,

1. Generate IDs for types as they appear in the code - We will create a map which we will fill as the types appear and their virtual tables are created. This will work well if are not doing separate compilation of files to binaries. However if we are doing separate compilation then in one file a type might have different ID than in other file leading to incorrect results at runtime.
2. Use only the name and its owner symbol table ID - We generate ID of types only using the name and its owner symbol table (we have access to both in the ASR even if the type is imported from somewhere else). This will be resilient if the hash function used to encode the combination of name and symbol table ID is resilient enough. 

cc: @certik

**Comment by certik** [comment]
I would start with a unique ID, then we don't need to do any kind of global tracking of types. Later we can improve upon it.

The unique ID of a type is generated by serializing it. We can start with our existing serializer, and later improve upon it to make it as short as possible. Since in general the serialization can be of arbitrary length, I would store it as a global variable (string, but allowing null values), probably a length + data. Then use a pointer to this. To compare if two types are the same, we need to compare the data of the two pointers, if they are the same, it is the same type.

**Comment by czgdp1807** [comment]
I guess what you are suggesting is `i8*` as type IDs. Well comparing two `i8*` at runtime is not very efficient (even for small strings one has to load the each i8 and then do integer compares). IMO, we can use `{i64, i8*}` as type IDs. `i64` contains the hash (generated at compile time because we know the serialised string for a type at compile time) of `i8*` string. Following are the cases,

**Case 1 - Hash of i8* doesn't match**

Then that's clear, type IDs don't match, and no need to run a loop to compare `i8*`.

**Case 2 - Hash of i8* matches**

Then we will have to run the loop to verify whether the content of `i8*` is also the same.

If we only use `i8*` all the times then we will have to run the loop all the times to compare two types at runtime. However with `{i64, i8*}` and with a nice low collision hash function, we will encounter the second case for non-equal types rarely.

> We can start with our existing serializer, and later improve upon it to make it as short as possible

Yes shorter strings can give us speedups but not as good as integer compares. And comparing hash functions will allow us to avoid those string comparisons for non-matching types.

**Comment by certik** [comment]
Yes, that sounds like a good plan. This should be fast enough, and I think this is only needed for the runtime type checking, which one shouldn't use in performance critical code anyway.

**Comment by czgdp1807** [comment]
True. Runtime polymorphism shouldn’t be used for HPC codes. Additionally Fortran’s runtime polymorphism is tricky because any class can become runtime polymorphic.
